src/graft_dag.cpp

```diff
@@ -0,0 +1,487 @@
+    std::vector<Bitset> host_nodes = host_dag_.GetSortedVectorOfNodeBitsets();
+    nodes.insert(nodes.end(), host_nodes.begin(), host_nodes.end());
+  }
+  std::sort(nodes.begin(), nodes.end());
+  return nodes;
+}
+
+BitsetVector GraftDAG::GetSortedVectorOfEdgeBitsets(bool include_host) const {
+  BitsetVector edges;
+  // Get graft edge bitsets.
+  for (const auto &idpair_idx_pair : graft_edges_) {
+    auto id_pair = idpair_idx_pair.first;
+    auto parent_bitset = GetDAGNode(id_pair.first)->GetBitset();
+    auto child_bitset = GetDAGNode(id_pair.second)->GetBitset();
+    Bitset edge_bitset = Bitset::PCSP(parent_bitset, child_bitset);
+    edges.push_back(edge_bitset);
+  }
+
+  // Get host edge bitsets.
+  if (include_host) {
+    BitsetVector host_edges = host_dag_.GetSortedVectorOfEdgeBitsets();
+    edges.insert(edges.end(), host_edges.begin(), host_edges.end());
+  }
+  std::sort(edges.begin(), edges.end());
+
+  return edges;
+}
+
+// ** Counts
+
+size_t GraftDAG::TaxonCount() const { return host_dag_.TaxonCount(); }
+
+size_t GraftDAG::NodeCount() const { return HostNodeCount() + GraftNodeCount(); }
+
+size_t GraftDAG::GraftNodeCount() const { return graft_nodes_.size(); }
+
+size_t GraftDAG::HostNodeCount() const { return host_dag_.NodeCount(); }
+
+size_t GraftDAG::EdgeCount() const { return HostEdgeCount() + GraftEdgeCount(); }
+
+size_t GraftDAG::GraftEdgeCount() const { return graft_edges_.size(); }
+
+size_t GraftDAG::HostEdgeCount() const {
+  return host_dag_.EdgeCountWithLeafSubsplits();
+}
+
+bool GraftDAG::IsNodeFromHost(size_t node_id) const {
+  return (node_id >= HostNodeCount());
+}
+
+bool GraftDAG::IsEdgeFromHost(size_t edge_id) const {
+  return (edge_id >= HostEdgeCount());
+}
+
+// ** Contains
+
+bool GraftDAG::ContainsGraftNode(const Bitset node_subsplit) const {
+  return subsplit_to_id_.find(node_subsplit) != subsplit_to_id_.end();
+}
+
+bool GraftDAG::ContainsGraftNode(const size_t node_id) const {
+  return (node_id >= HostNodeCount()) && (node_id < NodeCount());
+}
+
+bool GraftDAG::ContainsNode(const Bitset node_subsplit) const {
+  return host_dag_.ContainsNode(node_subsplit) || ContainsGraftNode(node_subsplit);
+}
+
+bool GraftDAG::ContainsNode(const size_t node_id) const {
+  return host_dag_.ContainsNode(node_id) || ContainsGraftNode(node_id);
+}
+
+bool GraftDAG::ContainsGraftEdge(const size_t parent_id, const size_t child_id) const {
+  return graft_edges_.find({parent_id, child_id}) != graft_edges_.end();
+}
+
+bool GraftDAG::ContainsEdge(const size_t parent_id, const size_t child_id) const {
+  return host_dag_.ContainsEdge(parent_id, child_id) ||
+         ContainsGraftEdge(parent_id, child_id);
+}
+
+// ** GP Operations
+
+QuartetHybridRequest GraftDAG::QuartetHybridRequestOf(const size_t parent_id,
+                                                      const bool left_clade_is_focal,
+                                                      const size_t child_id) const {
+  // Get nodes
```

Comment:
This isn't correct, as we discussed. I don't see a reason to merge it in.

Reply:
Removed.